gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/DiagnosticsManifestPlugin.java

```diff
@@ -0,0 +1,190 @@
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.api.tasks.TaskAction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class DiagnosticsManifestPlugin implements Plugin<Project> {
+
+    public static final String mergeDiagnosticsJson = "mergeDiagnosticsJson";
+
+    /**
+     * This plugin uses Gradle's Artifact Transforms to extract a single file from all the jars on the classpath
+     * (https://docs.gradle.org/current/userguide/artifact_transforms.html).
+     * All we do is tell gradle what function can turn 'usage=jar' into 'usage=(our thing)', and then declare a
+     * a view using 'usage=(our thing)' and it'll just run the {@link ExtractSingleFile} transform
+     * to process the jars and extract the stuff we want! Crucially, this is all cached _beautifully_.
+     */
+    @Override
+    public void apply(Project project) {
+        String fileToExtract = "sls-manifest/diagnostics.json";
+        String attribute = "extracted-" + fileToExtract;
+
+        project.getDependencies().registerTransform(ExtractSingleFile.class, details -> {
+            details.getParameters().getPathToExtract().set(fileToExtract);
+
+            // this USAGE_ATTRIBUTE is already present on everything, so gradle can figure out how to transform to our
+            // attribute value
+            details.getFrom()
+                    .attribute(Usage.USAGE_ATTRIBUTE, project.getObjects().named(Usage.class, Usage.JAVA_RUNTIME));
+            details.getTo()
+                    .attribute(Usage.USAGE_ATTRIBUTE, project.getObjects().named(Usage.class, attribute));
+
+            // these ones aren't really necessary, just for tidiness (seems bad to label something a jar when it's not)
+            details.getFrom()
+                    .attribute(
+                            LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE,
+                            project.getObjects().named(LibraryElements.class, LibraryElements.JAR));
+            details.getTo()
+                    .attribute(
+                            LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE,
+                            project.getObjects().named(LibraryElements.class, attribute));
+        });
+
+        // In order to get classes & resources from this project bundled into a jar, we make up this new
+        // configuration and add a 'self' dependency.
+        Configuration consumable = project.getConfigurations().create("runtimeClasspath2", conf -> {
+            conf.extendsFrom(project.getConfigurations().getByName("runtimeClasspath"));
+            conf.setDescription("DiagnosticsManifestPlugin uses this configuration to extract single file");
+            conf.setCanBeConsumed(true);
+            conf.setCanBeResolved(true);
+        });
+        project.getDependencies().add(consumable.getName(), project);
+
+        ArtifactView myView = consumable.getIncoming().artifactView(v -> {
```

Comment:
haven't played with artifactViews before but I'm pretty sure you can accomplish something similar by setting attributes on the configuration itself.

Either way, I think that this approach has a bit of a problem since it requires all project dependencies to be compiled, packaged into a JAR and then unpackaged to resolve the supported diagnostics. We've tried pretty hard to avoid doing so since `createManifest` is executed on `--write-locks` and compiling all dependant projects can be quiet expensive

Reply:
Ok have added a direct java-resources -> my attribute transform, so we now pass those tests that check the createManifest task doesn't compile stuff!!

As to your question about artifactView, I did actually try setting the attributes on the configuration directly, but couldn't seem to get it work work... have written a little essay in the comments to hopefully explain its existence. Maybe we can come back with greater understanding and delete it!!